### PR TITLE
Otp endpoints

### PR DESCRIPTION
**Pull request by spcheema**

### CODE REVIEW

File: src/Resource/Otp.php [hunk 1]
```diff
@@ -52,22 +52,58 @@ public function send(string $to, string $text, string $from = null, string $code
...
+     *
+     * @return array
+     * @throws AuthenticationException
+     * @throws InvalidPayloadException
+     * @throws ResourceNotFoundException
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function verfiyByRequestId(string $requestId, string $code)
+    {
+        $uri = $this->prepareApiUri($this->resourceUri . '/requestid/' . $requestId . '/validate');
```
Comment (dilantha) [inline src/Resource/Otp.php:68]: Lets validate `$requestId` to make sure the user can't send any malicious content.

File: src/Resource/Otp.php [hunk 2]
```diff
@@ -52,22 +52,58 @@ public function send(string $to, string $text, string $from = null, string $code
...
      * @throws InvalidPayloadException
      * @throws InvalidResponseException
      * @throws ResourceNotFoundException
      * @throws \GuzzleHttp\Exception\GuzzleException
      */
-    public function verfiyById(string $requestId, string $code)
+    public function verfiyByDestination(string $destination, string $code)
     {
-        $origin = !empty($from) ? $from : '';
+        $uri = $this->prepareApiUri($this->resourceUri . '/' . $destination . '/validate');
```
Comment (dilantha) [inline src/Resource/Otp.php:88]: Lets validate the destination number using something like phonenumberutil or other method to ensure they don't pass malicious strings.

File: src/Resource/Otp.php [hunk 3]
```diff
@@ -52,22 +52,58 @@ public function send(string $to, string $text, string $from = null, string $code
...
+     * @return array
+     * @throws AuthenticationException
+     * @throws InvalidPayloadException
+     * @throws InvalidResponseException
+     * @throws ResourceNotFoundException
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    private function verify($uri, $code)
+    {
         $jsonPayload = json_encode(compact('code'), JSON_FORCE_OBJECT);
```
Comment (dilantha) [inline src/Resource/Otp.php:107]: Lets validate the code is a numeric for the given length before passing it on

File: src/Resource/Otp.php [hunk 4]
```diff
@@ -95,43 +133,51 @@ public function verfiyById(string $requestId, string $code)
      * @throws ResourceNotFoundException
      * @throws \GuzzleHttp\Exception\GuzzleException
      */
-    public function getById(string $requestId): array
+    public function cancelByRequestId(string $requestId)
```
Comment (dilantha) [inline src/Resource/Otp.php:136]: Validate the `$requestId`

File: src/Resource/Otp.php [hunk 5]
```diff
@@ -95,43 +133,51 @@ public function verfiyById(string $requestId, string $code)
...
+     * @param string $destination Destination number
+     *
+     * @return array
      * @throws AuthenticationException
      * @throws InvalidResponseException
      * @throws ResourceNotFoundException
      * @throws \GuzzleHttp\Exception\GuzzleException
      */
-    public function cancelById(string $requestId)
+    public function cancelByDestination($destination)
```
Comment (dilantha) [inline src/Resource/Otp.php:154]: Validate destination

File: tests/Resource/OtpTest.php [hunk 6]
```diff
@@ -58,44 +58,23 @@ public function testCancelByIdNotFound(): void
         }
     }
 
-    public function testCancelByIdFailedWithConflict(): void
+    public function testCancelByRequestId(): void
     {
-        $this->expectException(ClientException::class);
-
-        $responseBody = '{"error":"The input code has already been verified."}';
+        $responseBody = '{"requestId":"404372541682577504482079","destination": "61474950859", "validUnitlTimestamp":"2020-11-06 13:59:11","createdTimestamp":"2020-11-06 13:49:11","lastEventTimestamp":"2020-11-06 13:49:33","status":"Cancelled"}';
```
Comment (dilantha) [inline tests/Resource/OtpTest.php]: Don't leave real numbers in public code

File: tests/Resource/OtpTest.php [hunk 7]
```diff
@@ -107,62 +86,59 @@ public function testCancelById(): void
...
-    public function testGetByIdNotFound(): void
+    public function testCancelByDestination(): void
     {
-        $this->expectException(ResourceNotFoundException::class);
+        $responseBody = '{"requestId":"404372541682577504482079","destination": "61474950859", "validUnitlTimestamp":"2020-11-06 13:59:11","createdTimestamp":"2020-11-06 13:49:11","lastEventTimestamp":"2020-11-06 13:49:33","status":"Cancelled"}';
 
         // Create a mock and queue two responses.
         $mock = new MockHandler([
-            new Response(404),
+            new Response(200, ['content-type' => 'application/json'], $responseBody),
```
Comment (dilantha) [inline tests/Resource/OtpTest.php:117]: Use a HTTP status ok constant

File: tests/Resource/OtpTest.php [hunk 8]
```diff
@@ -107,62 +86,59 @@ public function testCancelById(): void
...
+
         $handlerStack = HandlerStack::create($mock);
         $client = new Client(['handler' => $handlerStack]);
 
         try {
             $otp = new Otp($client);
-            $otp->getById('404372541682577504482079');
-        } catch (CredentialsException $e) {
-            $this->fail('This test should not have failed');
+            $otpResponse = $otp->cancelByDestination('61474950859');
```
Comment (dilantha) [inline tests/Resource/OtpTest.php]: Don't keep our number in public code, unless you like spam.

File: tests/Resource/OtpTest.php (hunk 7 above)
Comment (spcheema) [reply tests/Resource/OtpTest.php:117]: PHP/Guzzle does not have such constants. I dont mind adding a trait but then we have to look further for code coverage report.

**Pull request merged**